server/raft.go

```diff
@@ -4332,25 +4333,22 @@ func (n *raft) switchToLeader() {
 
 	n.lxfer = false
 	n.updateLeader(n.id)
-	n.switchState(Leader)
-
-	// Wait for messages to be applied if we've stored more, otherwise signal immediately.
-	if n.pindex > n.applied {
-		n.aflr = n.pindex
-	} else {
-		n.signalAppliedFloor()
+	leadChange := n.switchState(Leader)
+
+	if leadChange {
+		// Wait for messages to be applied if we've stored more, otherwise signal immediately.
+		// It's important to wait signaling we're leader if we're not up-to-date yet, as that
+		// would mean we're in a consistent state compared with the previous leader.
+		if n.pindex > n.applied {
+			n.aflr = n.pindex
+		} else {
+			n.aflr = 0
```

Comment:
Not entirely clear what 0 means here, I know the comment above says signalling disabled, but is that implying this doesn't track upwards with applied always?

Reply:
We only need to track if we have a log with entries that are not applied. If all our entries are applied we can signal immediately. Resetting `n.aflr` is just for sanity so we couldn't signal leader twice.

Have added this comment:
```go
// We know we have applied all entries in our log and can signal immediately.
// For sanity reset applied floor back down to 0, so we aren't able to signal twice.
```